lib/credentials/token_file_web_identity_credentials.js

```diff
@@ -0,0 +1,149 @@
+   *  AWS.config.credentials = new AWS.TokenFileWebIdentityCredentials(
+   *   // optionally provide configuration to apply to the underlying AWS.STS service client
+   *   // if configuration is not provided, then configuration will be pulled from AWS.config
+   *   {
+   *     // specify timeout options
+   *     httpOptions: {
+   *       timeout: 100
+   *     }
+   *   });
+   * @see AWS.Config
+   */
+  constructor: function TokenFileWebIdentityCredentials(clientConfig) {
+    AWS.Credentials.call(this);
+    this.clientConfig = clientConfig;
+    this.get(function() {});
+  },
+
+  /**
+   * Returns params from environment variables
+   *
+   * @api private
+   */
+  getParamsFromEnv: function getParamsFromEnv() {
+    var ENV_TOKEN_FILE = 'AWS_WEB_IDENTITY_TOKEN_FILE',
+        ENV_ROLE_ARN = 'AWS_IAM_ROLE_ARN';
+    if (process.env[ENV_TOKEN_FILE] && process.env[ENV_ROLE_ARN]) {
+      return {
+        envTokenFile: process.env[ENV_TOKEN_FILE],
+        roleArn: process.env[ENV_ROLE_ARN],
+        roleSessionName: process.env['AWS_IAM_ROLE_SESSION_NAME']
+      };
+    }
+  },
+
+  /**
+   * Returns params from shared config variables
+   *
+   * @api private
+   */
+  getParamsFromSharedConfig: function getParamsFromSharedConfig() {
+    try {
+      var profiles = AWS.util.getProfilesFromSharedConfig(iniLoader);
+      var profile = profiles[process.env.AWS_PROFILE || AWS.util.defaultProfile] || {};
+
+      if (Object.keys(profile).length === 0) {
+        throw AWS.util.error(
+          new Error('Profile ' + this.profile + ' not found'),
+          { code: 'TokenFileWebIdentityCredentialsProviderFailure' }
+        );
+      }
+
+      return {
+        envTokenFile: profile['web_identity_token_file'],
+        roleArn: profile['role_arn'],
+        roleSessionName: profile['role_session_name']
+      };
+    } catch (err) {
+      throw err;
+    }
+  },
+
+  /**
+   * Refreshes credentials using {AWS.STS.assumeRoleWithWebIdentity}
+   *
+   * @callback callback function(err)
+   *   Called when the STS service responds (or fails). When
+   *   this callback is called with no error, it means that the credentials
+   *   information has been loaded into the object (as the `accessKeyId`,
+   *   `secretAccessKey`, and `sessionToken` properties).
+   *   @param err [Error] if an error occurred, this value will be filled
+   * @see AWS.Credentials.get
+   */
+  refresh: function refresh(callback) {
+    this.coalesceRefresh(callback || AWS.util.fn.callback);
+  },
+
+  /**
+   * @api private
+   */
+  load: function load(callback) {
+    var self = this;
+    try {
+      var params = self.getParamsFromEnv();
+      if (!params) {
+        params = self.getParamsFromSharedConfig();
+      }
+      if (params) {
+        var oidcToken = fs.readFileSync(params.envTokenFile, {encoding: 'ascii'});
+        if (self.webIdentityCredentials) {
+          self.webIdentityCredentials.service.config.params.WebIdentityToken = oidcToken;
```

Comment:
The `self.webIdentityCredentials.service` is loaded [lazily](https://github.com/aws/aws-sdk-js/blob/a0010e6fde102c86496ecd9683f50554f3493c3f/lib/credentials/web_identity_credentials.js#L93). It's better to check if `self.webIdentityCredentials.service` is defined. If not we need to call the [`createClients`](https://github.com/aws/aws-sdk-js/blob/a0010e6fde102c86496ecd9683f50554f3493c3f/lib/credentials/web_identity_credentials.js#L107)

Reply:
Resolved in https://github.com/aws/aws-sdk-js/pull/2737/commits/c0c197a39ec566a7d449390fdbbbcd1275e41b4d